### DISCUSSION TITLE
Conveyor as a signing tool?

### DISCUSSION THREAD

**Discussion by ennerf** [hydraulic-software/conveyor#46] [Ideas]
I recently ran into some issues where graal native images generated on a mac can't easily be executed on another machine due to lack of signing. Since the standard signing process is too painful, I was considering setting up a wrapper package just to get the executable signed...

Given that all the functionality is already there, would it be possible to add signing as a feature to the command line tool?

e.g. `conveyor sign <executable/lib> <id?>`

**Comment by mikehearn** [comment]
I think standalone Mac binaries outside a bundle need a slightly different signing approach (to embed an Info.plist), but it might be possible to do that. I don't recall if they work if you extract them from their bundle; worth testing! For single unsigned binaries the user can use "ad-hoc signing" to whitelist them for use on their specific ARM Mac. The user has to run something like `codesign -s - ./binary` and then it should be accepted by their kernel. This is how Xcode is able to run code that was just compiled, so it could be put into a `curl | bash` type script.

For now the workaround is to do as you suggest - just have a simple config that imports the binary to be signed, use `conveyor make windows-app` or `conveyor make notarized-mac-app`, and grab the result from the `output` directory. You'll have to disable updates on macOS (`app.updates = none`) as otherwise Conveyor will inject a dependency on libconveyor, which is used to start Sparkle. I don't think we test the no updates codepath much though. It'll get a bit more convenient in the next major release because for various reasons the CLI will become able to build multiple tasks at once, and Conveyor will render simple icons for you if you don't have any, so the work required to bring up a minimal app will shrink again.

Originally the design idea was to have each task in the build be a separate sub-command that could be invoked via the CLI or via a generic build configuration, but it wasn't clear how to make that work w.r.t. licensing. If there's no site URL there's no definition of a project.

One possibility for making CLI apps easier to distribute that we've been researching lately is sandboxing. CLI apps are relatively "easy" to sandbox using e.g. the Chromium sandbox. If the app consists only of CLI apps then we could potentially run Conveyor as a service - you upload unsigned binaries, we give back a URL to a download site for packages signed with our keys that runs the provided code inside the sandbox. You wouldn't need to own signing keys at all, just a Conveyor license (or be open source). In case of a security bug being found in the sandbox we can immediately kick off updates to patch it.

**Reply by ennerf** [reply]
I just did some tests with binaries created on CI, packaged/signed on Windows, and executed on another Mac. It looks like the culprit wasn't the signing, but the quarantine flag. 

`sudo xattr -r -d com.apple.quarantine file` made the unsigned version work.

Uploading an extracted/signed executable acts exactly the same as the unsigned version.

**Reply by mikehearn** [reply]
There may be a difference in behavior between ARM and Intel Macs in this regard. ARM Macs are stricter about signing.
